### DISCUSSION TITLE
Can't Get Door Locks to Work

### DISCUSSION THREAD

**Discussion by PeteCal** [zwave-js/zwave-js-ui#2633] [Q&A]
I am not sure this is a question for this forum.  If it isn't, maybe someone can point me in the right direction.

I have three door locks, all different manufactures.  I get the same "Values" for all of them.  See the image.  Note, I also have a zwave thermostat and it works great with zwavejs2mqtt.

I have been using these locks for years with OpenHAB.  In fact, if I swap the SD card (Raspberry Pi) back to the OpenHAB SD, they still work.  When I swap back to the zwavejs2mqtt SD I get the image.

Details:
I am using an AEOTEC stick. 
I included all the devices when using OpenHAB.
If I click the arrow> for the Target lock mode I see a flurry of activity in MQTT Explorer that seems to send all 15 topics on the Values page (only 6 or 9 are in the image, the rest are lower down).

What could be my problem?

![Node28Zwavejs2MQTT](https://user-images.githubusercontent.com/33606511/188718941-5fe55c55-1255-44ed-b4b1-a4d85f783ca5.png)

**Comment by robertsLando** [comment]
Did you tried to re-interview your door locks? Anyway we need driver logs to help you here.

**Reply by robertsLando** [reply]
cc @AlCalzone

**Comment by AlCalzone** [comment]
> I have been using these locks for years with OpenHAB. In fact, if I swap the SD card (Raspberry Pi) back to the OpenHAB SD, they still work. When I swap back to the zwavejs2mqtt SD I get the image.

Have you configured security keys in zwavejs2mqtt to be the same as in OpenHAB?

**Reply by PeteCal** [reply]
There is a high probability the security keys are the problem.  But in OpenHAB I can find only one "network security key" but zwavejs2mqtt has four.   I am reserching what that means now.

**Reply by AlCalzone** [reply]
That's the one for S0. Generate the other ones though, you'll need them for newer devices.

**Reply by PeteCal** [reply]
Thanks.  I set that for S0 but it doesn't seem to have helped, even after an hour or more.
Here is a Debug log.  At about line 280 I did a refresh of node 28 and then another refresh.

Of course I don't know how to interperate it but I do see "is alive, ready to be used, ping successful"  But the refresh times out?
[zwavejs2mqttDebugLog.txt](https://github.com/zwave-js/zwavejs2mqtt/files/9509155/zwavejs2mqttDebugLog.txt)

**Reply by AlCalzone** [reply]
1. you're a bit behind with updates: `DRIVER version 9.5.0`
2. Please make a [driver log](https://zwave-js.github.io/zwavejs2mqtt/#/troubleshooting/generating-logs?id=driver-logs), loglevel **debug** and attach it here as a file. Your log was on a different loglevel. Preferably download the logfile via the store, this preserves formatting and makes it MUCH easier to read.

**Reply by PeteCal** [reply]
I believe I found the file you want.  The procedure didn't match what was on the github README page.

[zwavejs_2022-09-08.log](https://github.com/zwave-js/zwavejs2mqtt/files/9528495/zwavejs_2022-09-08.log)

As for the version, that could be a problem,  When I first tried to install z2m I had a great deal of difficuility.  I tried a few different methods, snap, docker, etc.  I spent many hours trying over and over then suddenly it was there.  I was using another SD card not the one I was developing on.  When I got it to work I worked to get it installed on my development SD.  It took a few attempts but finally it showed up.  When I tried to use it I couldn't get it to work.  So I switched back to the SD that had a working version and transferred my development files to that SD.  I'm not up to going through that again if I can avoid it.  What do you think?  Can I limp with 9.5.0?

**Reply by AlCalzone** [reply]
> I believe I found the file you want. The procedure didn't match what was on the github README page.

In which way was it different?

**Reply by AlCalzone** [reply]
Anyways, there's nothing in that log that shows any interaction with your door lock. Please re-interview it, then try to control it, so I can see what is happening.

**Reply by PeteCal** [reply]
> > I believe I found the file you want. The procedure didn't match what was on the github README page.
> 
> In which way was it different?

No big deal but the animation on:
https://zwave-js.github.io/zwavejs2mqtt/#/troubleshooting/generating-logs?id=download-zip
shows an item "logs" with the log files as sub items below it.   My installation doesn't have the item "logs", the log and json files just start after the last folder which is "sessions".
Like I said, no big deal but it just gave me a little doubt.
Now, at around 8:45 PM US CT I did the re-interview of node 28 and changed **[28-98-0-targetMode] Target lock mode to [0] Unsecured** and clicked the right facing arrow twice a few minutes apart.  I then did REFRESH twice.  
It looks like I have to wait for zwavejs_current.log to go to completion in order to download it.  Note, MQTT Explorer reports that a topic like that appeared 24 minutes earlier.  I had nothing to do with that.  I wasn't even in the same room.
When the current log closes I will download it and forward it.

**Reply by AlCalzone** [reply]
> Like I said, no big deal but it just gave me a little doubt.

Ah, yeah this changed in a recent version. Like I said earlier, you're a bit behind.

**Reply by PeteCal** [reply]
Well, All I have seems to be a mess.  I sent 09-08 earlier.  Here is 09-09.  I doubt it is of use.
[zwavejs2mqtt_2022-09-09.log](https://github.com/zwave-js/zwavejs2mqtt/files/9537225/zwavejs2mqtt_2022-09-09.log)

I also went back and looked at previous logs.  Almost nothing. See below.

[zwavejs2mqtt_2022-09-05.log](https://github.com/zwave-js/zwavejs2mqtt/files/9537239/zwavejs2mqtt_2022-09-05.log)
[zwavejs2mqtt_2022-09-06.log](https://github.com/zwave-js/zwavejs2mqtt/files/9537242/zwavejs2mqtt_2022-09-06.log)
[zwavejs2mqtt_2022-09-07.log](https://github.com/zwave-js/zwavejs2mqtt/files/9537243/zwavejs2mqtt_2022-09-07.log)

Ii think you will tell me to get the latest version.  I cringe at the thought.  But I guess I better clone the SD card I have, then try for the latest.

**Reply by AlCalzone** [reply]
And those are not the correct logs ;) You downloaded the application logs, not the driver logs.

Honestly I haven't heard of anyone having these kinds of issues with upgrading. You might want to try not running it barebones, e.g. using [Docker](https://zwave-js.github.io/zwavejs2mqtt/#/getting-started/docker), or [Snap](https://snapcraft.io/install/zwavejs2mqtt/raspbian)

**Reply by PeteCal** [reply]
Sorry about the logs.  I got in the wrong section.  Nothing of interest beyond the 09-08 one.

Last time I tried both snap and docker.  Both gave me problems.  As I said, at some point I get it working but I don't know what I did.  Then I tried to install it on another SD and couldn't get that to work so II went back to the working SD and moved all my files over to it.

I'm  thinking that in all my bumbling around I have some kind of crossed up system.  I wish there was some way to "uninstall" so I could start with a blank slate,  For now I guess I'll start with Docker.

**Reply by PeteCal** [reply]
Just to let youu know, I spent half an hour or so trying to use docker.  I had to:
 docker kill bf63016ada22 to get rid of some stuff.  Then I tried both the "volumes as persistance" and the "Using local folder as persistance".  Both gave me the same problem.  I just could not come up with the right way to follow the Note about replacing the "insert_stick .ref........".  Z2M was using /dev/ttyACMO so I kept trying that but I must have not replace the text correctly.
So I tried "as a service".  That gave me some kind of container conflict.  Based on my previous attempts this kind of two steps forward, three steps back will go on forever.  I guess I was right whe I said I wish I could uninstall Z2M.  A lot of searching doesn't come up with a method.

I will try snap now.....

**Reply by PeteCal** [reply]
I went through the snap steps and got to exactly the "Raspberry Pi users running Raspbian/Debian, read this thread" problem.  The chosen solution by jmgiaver didn't work for me.  I got a whole hosts of errors,........ignored.
I'll try the NodeJS approach.  I know it is described as "the most complex way" but I feel more comfortable with linux command lines.

**Reply by PeteCal** [reply]
Nope.  All went well until I tried ./zwavejs2mqtt I got:
-bash: ./zwavejs2mqtt: cannot execute binary file: Exec format error

I don't think I can ask you to unravel the mess I am in.  It is just too complex.
Do you have any suggestion(s) where I should go from here?  And I hope it doesn't end in "....LL"  ;-)

**Reply by AlCalzone** [reply]
Tbh, I'm just writing the library that other projects including zwavejs2mqtt are using. All I can do is give pointers. Like this one:

> /dev/ttyACMO

The "O" should be a zero.

**Reply by PeteCal** [reply]
Ouch, I knew that from long ago but forgot.
Now I ran:
 docker run --rm -it -p 8091:8091 -p 3000:3000 --device=/dev/ttyACM0:/dev/zwave \--mount source=zwavejs2mqtt,target=/usr/src/app/store zwavejs/zwavejs2mqtt:latest

The first time it said the port was allocated so I had to kill z2m using the container id.
Now if I enter the same line I get the  logo then one line of:
2022-09-10 01:20:28.290 WARN STORE: scenes.json not found

Then, this log entry about once every 20 or 25 seconds:
```

022-09-10T00:55:52.629Z DRIVER   version 9.5.0
2022-09-10T00:55:52.630Z DRIVER   
2022-09-10T00:55:52.631Z DRIVER   starting driver...
2022-09-10T00:55:52.646Z DRIVER   opening serial port /dev/ttyACM0
2022-09-10T00:56:01.679Z DRIVER   Failed to open the serial port: Error: No such file or directory, cannot open 
                                  /dev/ttyACM0
2022-09-10T00:56:01.681Z DRIVER   destroying driver instance...
2022-09-10T00:56:01.684Z DRIVER   driver instance destroyed
2022-09-10T00:56:04.700Z DRIVER  
```
When I open 192.168.1.115:8091 and view the zwave settings, the Serial Port is /dev/ttyACM0 just like it was with the older working version of z2m but now it says the port fails to open. And the driver version is still 9.5.0

I think I am going to compare files in the working SD to this one and see what might be different.

**Reply by kpine** [reply]
> --device=/dev/ttyACM0:/dev/zwave 

With this configuration the path in the container is `/dev/zwave` **not** `/dev/ttyACM0`. Pay close attention to the minimum required settings in the docs: https://zwave-js.github.io/zwavejs2mqtt/#/getting-started/quick-start?id=minimum-settings

**Reply by PeteCal** [reply]
I had a second thought.  Uninstall docker and start over by installing docker and then the docker run...... as above.  Almost the same result except in zwave the port was "/dev/zwave " not /dev/tty/ACM0.  I changed  it and saved it but still the same error message as above except the DRIVER version is now 10.0.4

**Reply by kpine** [reply]
You still need to enter the port path and save it. It's just a pre-filled text box, not set.

**Reply by PeteCal** [reply]
Yes, I did make it /dev/ttyACM0 and save it.

**Reply by kpine** [reply]
Except the setting is **not** `/dev/ttyACM0`, it's `/dev/zwave`.

`--device=/dev/ttyACM0:/dev/zwave` -> host path is `/dev/ttyACM0`, container path is `/dev/zwave`. zwavejs2mqtt runs in the container.

**Comment by PeteCal** [comment]
**kpine:**
That very note has been fighting me all day.  It says
"Replace /dev/serial/by-id/insert_stick_reference_here (only the first half of the X:Y mapping!) with the serial device for your Z-Wave stick".
The highlighted text is:
"**/dev/serial/by-id/insert_stick_reference_here**" so that was what I tired to replace.  Thus I left, **":/dev/zwave "** in place.  
I have no idea what " **(only the first half of the X:Y mapping!)** " means and neither does google.
So please tell me what that line should look like.  I tried many variations that were rejected but the one I used was accepted.

**Reply by robertsLando** [reply]
In UI settings just use `/dev/zwave`. In docker-compose just use `/dev/ttyACM0:/dev/zwave`

**Reply by PeteCal** [reply]
Under settings UI in my installation, I have only Dark mode and Use tabs for navigation.  No place for  /dev/zwave.  
The docker-compose.yml I used from that site already had the device filled in correctly.

**Reply by kpine** [reply]
The USB path is in the Z-Wave section.

The docker compose file doesn't configure zwavejs2mqtt, it only adds the USB device into the container. You still need to configure the USB path in zwavejs2mqtt itself, in the Z-Wave config section. The only way you can avoid this is by manually creating a settings.json file. The USB path is not configured by default.

**Reply by PeteCal** [reply]
Since my zwave thermostat is working, doesn't that mean the settings are correct?

FYI I went back to my development SD with version 9.5.0.  The thermostat works fine but I can't access any of the three locks and the S0 key is the same for both SD cards.

**Reply by AlCalzone** [reply]
Now I'll see driver logs of the failed communication attempts. Loglevel debug, attach here as a file using drag & drop.

**Reply by PeteCal** [reply]
[zwavejs_2022-09-12.log](https://github.com/zwave-js/zwavejs2mqtt/files/9551576/zwavejs_2022-09-12.log)

Hot off the pressess
Node 28 shows a green check under security
No, it doesn't anymore.  It did 10 minutes ago

**Reply by AlCalzone** [reply]
No response when querying the secure commands, so the driver deduces that the lock isn't secure.
To get this back (or at least try to), re-interview again and check the box "reset security classes" so this is figured out again (hopefully).

Also: upgrade :) The latest version does not simply forget that a lock is secure if it has worked securely before.

**Reply by PeteCal** [reply]
Well, I went through an upgrade process but I don't know if it worked.
I have been determining the version from the log file but the newest one I have is the one I sent you.  Is there another way to get the version?

**Reply by PeteCal** [reply]
My bad.  I see:
zwavejs2mqtt 7.1.0
zwavejs 10.0.4

**Reply by PeteCal** [reply]
I was watching the icons under "Security" for the three locks, Nodes 26, 28, 32. They seem to change at rendom from ? to Green, to Red.
I considered the possibility that the security keys were wrong for S2 Unauthenticated, S2 Authenticated, S2 Access Control .  They were the same for S0 Legacy because I supplied them.
I went back to the SD  with the minimum system.  All the locks still respone.  I copied the other security keys, then switched to the development system and put in the security keys that were used by the minimum SD.   That did not help.  Same as before, the locks fields are "undefined".
Here is the fresh log file in case you want to take a look at it:
[zwavejs_2022-09-13.log](https://github.com/zwave-js/zwavejs2mqtt/files/9560108/zwavejs_2022-09-13.log)

Unless you spot something I think I will have to go to the nuclear option.  Start with the minimum SD and incrementally add the programs and files I need to recreate the development system.  Maybe 20 or 30 of them.  I say incrementally because I will add a few then check that z2m is still accessing the locks, then add more.

Can you think of anything else short of this complete rebuild?

Thanks for the help

**Reply by AlCalzone** [reply]
It's definitely not an issue with the host system. Z-Wave communication clearly works, but the lock isn't responding to secure commands. It's not the S2 keys either, the lock is either included with S0 or doesn't support S2 - impossible to distinguish after inclusion.

Right at the start, there's one exchange that makes me think your S0 key is wrong:
```
2022-09-13T05:28:06.857Z DRIVER « [Node 028] [REQ] [ApplicationCommand]
                                  └─[SecurityCCNonceGet]
...
2022-09-13T05:28:07.273Z DRIVER » [Node 028] [REQ] [SendData]
                                  │ transmit options: 0x05
                                  │ callback id:      22
                                  └─[SecurityCCNonceReport]
                                      nonce: 0xcd024f7fd4e658f4
...
2022-09-13T05:28:07.479Z DRIVER « [Node 028] [REQ] [ApplicationCommand]
                                  └─[SecurityCCCommandEncapsulation] [INVALID]
                                      error: Invalid auth code, won't accept security encapsulated command.
```

In plain english: Lock requests a nonce (temporary value for encryption), Z-Wave JS responds, Lock uses nonce to encrypt command, sends it. Z-Wave JS is unable to decrypt. The only thing that could be wrong at this point is the S0 key, which also explains why the lock won't respond to any secure commands - it cannot decrypt them either.

**Reply by AlCalzone** [reply]
If you want to go "nuclear", exclude the locks, then factory reset them to get rid of any stored keys, then re-include them, so they get the key you've actually configured.

**Reply by PeteCal** [reply]
Al, you nailed it.  Slam dunk.  I had screwed up the security code.  One character out of 32.  I put in a "D" where I should have put "0".  I was getting it from a blury screen capture from the program that used the stick before.  But when I created the minimal system I put the key in correctly so that worked.  So, finally after getting all the real important stuff straightened out, it was this little thing. 

I was an excruciating journey since July 2 when I started with Z2M.  I know I did a lot of things wrong but as I look back I think one of the things that cost me a lot of time and tons of frustration was the "Note" in the image. 

![Z2MNote](https://user-images.githubusercontent.com/33606511/190289629-de003ca9-de11-4131-b384-3e632748447d.jpg)

 I tried over and over to get my /dev/ttyACMO entered into the line correctly.  In the end, I never did succeed.  The other website I used had it filled in correctly so that is why the install exceeded when I couldn't make that happen based on the Note.  Maybe it would help to provide a correct example.  The same applies to the Run as a service instructions.

Also, I couldn't get the NodeJS or PKG version to work because I always ended with some Exec error.  I tried to search that out but it is caused by many many things.  I was going nowhere.

All in all it looks like Z2M will be my solution.

Thanks Al, and the rest for getting me through this.

Pete

**Reply by kpine** [reply]
Regarding the note, are you saying you don't understand how find the USB path? If so, all you need to do is list the files in the directory:

```shell
$ ls -l /dev/serial/by-id
total 0
lrwxrwxrwx 1 root root 13 Jul  5 18:09 usb-Silicon_Labs_CP2102N_USB_to_UART_Bridge_Controller_bae21b797a3eeb11a8c12ebcf28b3f46-if00-port0 -> ../../ttyUSB0
```

Therefore _my_ controller USB path is `/dev/serial/by-id/usb-Silicon_Labs_CP2102N_USB_to_UART_Bridge_Controller_bae21b797a3eeb11a8c12ebcf28b3f46-if00-port0`. It's not feasible to provide a "correct" example path that you can just copy and paste, because with the introduction of 700-series controllers and later, the path includes a serial number which is unique.

**Reply by PeteCal** [reply]
No, it is mush simlper then that.  Where in the string:
 --device=/dev/serial/by-id/insert_stick_reference_here:/dev/zwave \
do I put  /dev/ttyACMO.
Just after the "=" and remove or keep the rest of the string?
Just after the "by-id/" and remove or keep the rest of the string?
In place of "insert_stick_reference_here"?
Do I keep the ":"?
I tried every thing I could think of.

Similarly, in docker-compose.yml 
"- '/dev/serial/by-id/insert_stick_reference_here:/dev/zwave'" presents a similar problem.
The site I used just said, " - '/dev/ttyACM0:/dev/ttyACM0'  " and it worked.  Maybe you folks are a few levels above me in this.

**Comment by PeteCal** [comment]
Success!!   After many hours over three days I finally cracked the problem.  I  continued trying to install using the instructions in the Readme but it kept acting like it couldn't get to the AEOTEC stick.  I developed the theory that there could be old code on the SD Card from other programs that used the stick.  So I decided to try with a minimum system.  I installed the minimum RPi OS and installed only Mosquitto and z2m which required Docker also.   With that I returned to using the Readme but kept getting the same problem.  I did some searching and found a site that had a different method of installing z2m.   I tried that and, after more then a typical number of dead ends and retries, it worked!

I can now reach all three door locks and the thermostat is fine.  The locks report the  state of the bolt correctly.  Now I have to decipher all the other commands that show up.

Now my alternatives are (1) go back to my development SD and see if I can get z2m to work using the method on  the new site or (2) use the minimum SD and transfer 20 or 30 my development files to the minimum system.  I think I will stat with #1.

Thanks folks, for all the help and sticking with me.

Pete
